crates/node-maintainer/src/graph.rs

```diff
@@ -219,62 +230,30 @@ impl Graph {
             q.extend(self.inner[node].children.values());
         }
 
-        // Verify that direct graph makes all dependencies available to
-        // dependents.
-        for edge_idx in self.inner.edge_indices() {
-            let (dependent, dependency) = self
-                .inner
-                .edge_endpoints(edge_idx)
-                .ok_or(GraphValidationError(format!("Missing edge: {edge_idx:?}")))?;
-
-            let dependent = &self.inner[dependent];
-            let dependency = &self.inner[dependency];
-
-            let edge = self
-                .inner
-                .edge_weight(edge_idx)
-                .ok_or(GraphValidationError(format!(
-                    "Missing edge weight: {edge_idx:?}"
-                )))?;
-
-            let dependency_parent = dependency.parent.ok_or(GraphValidationError(format!(
-                "Missing dependency parent: {:?}",
-                dependent.package.resolved(),
-            )))?;
-
-            // Check parent->child relationship
-            let dependency_name = UniCase::new(dependency.package.name().into());
-            if !self.inner[dependency_parent]
-                .children
-                .contains_key(&dependency_name)
-            {
-                return Err(GraphValidationError(format!(
-                    "Dependency {:?} is not in the children of {:?}",
-                    dependency.package.resolved(),
-                    self.inner[dependency_parent]
-                )));
-            }
-
-            // Parent of the dependency should be an ancestor of the dependent
-            // or dependency should be in dependent's subtree.
-            if !self.is_ancestor(dependent.idx, dependency.idx)
-                && !self.is_ancestor(dependency_parent, dependent.idx)
-            {
-                return Err(GraphValidationError(format!(
-                    "Dependency {:?} is unreachable from {:?}",
-                    dependency.package.resolved(),
-                    dependent.package.resolved(),
-                )));
-            }
-
-            // Check that dependency satisfies the requirement
-            if !dependency.package.resolved().satisfies(&edge.requested)? {
-                return Err(GraphValidationError(format!(
-                    "Dependency {:?} does not satisfy requirement {:?} from {:?}",
-                    dependency.package.resolved(),
-                    edge.requested,
-                    dependent.package.resolved(),
-                )));
+        // Verify that depencies are satisfied by the logical hierarchy.
+        for dependent in self.inner.node_weights() {
+            for (dep_name, edge_idx) in &dependent.dependencies {
+                let edge = &self.inner[*edge_idx];
+
+                if let Some(dep_idx) = self.resolve_dep(dependent.idx, dep_name) {
```

Comment:
...and replaced it with a bunch of `resolve_dep`. Basically each dependency should resolve to correct version. Easy to ready, easy to debug!